pageserver/src/tenant/timeline.rs

```diff
@@ -1602,350 +1599,169 @@ impl Timeline {
-            .as_ref()
-            .ok_or_else(|| anyhow!("cannot download without remote storage"))?;
-
-        let disk_consistent_lsn = up_to_date_metadata.disk_consistent_lsn();
+        })
+        .await
+        .map_err(anyhow::Error::new)
+        .and_then(|x| x)
+        .with_context(|| {
+            format!(
+                "process tenant {} timeline {} directory",
+                self.tenant_id, self.timeline_id
+            )
+        })?;
 
-        let local_layers = {
-            let guard = self.layers.read().await;
-            let layers = guard.layer_map();
-            layers
-                .iter_historic_layers()
-                .map(|l| (l.filename(), guard.get_from_desc(&l)))
-                .collect::<HashMap<_, _>>()
-        };
+        let num_layers = loaded_layers.len();
 
-        // If no writes happen, new branches do not have any layers, only the metadata file.
-        let has_local_layers = !local_layers.is_empty();
-        let local_only_layers = match index_part {
-            Some(index_part) => {
-                info!(
-                    "initializing upload queue from remote index with {} layer files",
-                    index_part.layer_metadata.len()
-                );
-                remote_client.init_upload_queue(index_part)?;
-                self.create_remote_layers(index_part, local_layers, disk_consistent_lsn)
-                    .await?
-            }
-            None => {
-                info!("initializing upload queue as empty");
-                remote_client.init_upload_queue_for_empty_remote(up_to_date_metadata)?;
-                local_layers
-            }
-        };
+        guard.initialize_local_layers(loaded_layers, disk_consistent_lsn + 1);
 
-        if has_local_layers {
-            // Are there local files that don't exist remotely? Schedule uploads for them.
-            // Local timeline metadata will get uploaded to remove along witht he layers.
-            for (layer_name, layer) in &local_only_layers {
-                // XXX solve this in the type system
-                let layer_path = layer
-                    .local_path()
-                    .expect("local_only_layers only contains local layers");
-                let layer_size = layer_path
-                    .metadata()
-                    .with_context(|| format!("failed to get file {layer_path:?} metadata"))?
-                    .len();
-                info!("scheduling {layer_path:?} for upload");
-                remote_client
-                    .schedule_layer_file_upload(layer_name, &LayerFileMetadata::new(layer_size))?;
+        if let Some(rtc) = self.remote_client.as_ref() {
+            if !needs_upload.is_empty() {
+                for (layer, m) in needs_upload {
+                    rtc.schedule_layer_file_upload(&layer.layer_desc().filename(), &m)?;
+                }
+                rtc.schedule_index_upload_for_file_changes()?;
             }
```

Comment:
Else the caller has already done `remote_client.schedule_index_upload_for_metadata_change(up_to_date_metadata)` on line tenant.rs:460 or so.

Reply:
I guess this comment is obsolete?

Reply:
No it was just an observation I added to help you.